File: debop4s-data/src/test/scala/com/github/debop4s/data/tests/mapping/parallel/JpaParallelsTest.scala
package com.github.debop4s.data.tests.mapping.parallel

import com.github.debop4s.core.utils.{ToStringHelper, Hashs}
import com.github.debop4s.data.jpa.utils.{JpaParCallable, JpaParRunnable, JpaParallels}
import com.github.debop4s.data.model.HibernateEntity
import com.github.debop4s.data.tests.AbstractJpaTest
import java.{util, lang}
import javax.persistence._
import org.hibernate.Hibernate
import org.hibernate.{annotations => hba}
import org.junit.{Before, Test}
import org.slf4j.LoggerFactory
import org.springframework.transaction.annotation.Transactional
import scala.collection.mutable.ArrayBuffer
import org.hibernate.annotations.LazyCollectionOption

/**
 * 테스트 시 RDMBS의 Connection 수를 많이 늘리던가 entityCount 수를 줄여야 합니다.
 * Created by debop on 2014. 3. 10.
 */
@Transactional
class JpaParallelsTest extends AbstractJpaTest {

    private lazy val log = LoggerFactory.getLogger(getClass)

    @PersistenceContext val em: EntityManager = null

    val entityCount = 100

    @Before
    def before() {
        // 캐시는 삭제되지 않는다!
        em.createQuery("delete from ParallelOrderItem").executeUpdate()
        em.createQuery("delete from ParallelOrder").executeUpdate()
        em.flush()
        em.clear()
    }

    @Test
    //    @Transactional(propagation = Propagation.REQUIRES_NEW)
    //    @Rollback(false)
    def parallelSaveAndRead() {

        val orderIds = ArrayBuffer[Long]()

        JpaParallels.run(emf, (0 until entityCount).toIterable) { (em, x) =>
            val order = createOrder(x)
            em.persist(order)
            orderIds += order.id
        }

        Thread.sleep(10)

        var orders =
            JpaParallels.call(emf, orderIds) { (em, x) =>
                val order = em.find(classOf[ParallelOrder], x)
                Hibernate.initialize(order.items)
                em.detach(order)
                order
            }

        assert(orders.forall(x => x != null))

        // 여기서부터는 hibernate-redis second 캐시에서도 잘 되는지 알아보기 위함입니다.
        orders =
            JpaParallels.call(emf, orderIds) { (em, x) =>
                val order = em.find(classOf[ParallelOrder], x)
                Hibernate.initialize(order.items)
                em.detach(order)
                order
            }

        assert(orders.forall(x => x != null))

        orders =
            JpaParallels.call(emf, orderIds) { (em, x) =>
                val order = em.find(classOf[ParallelOrder], x)
                Hibernate.initialize(order.items)
                em.detach(order)
                order
            }

        assert(orders.forall(x => x != null))
    }

    @Test
    //@Transactional(propagation = Propagation.REQUIRES_NEW)
    //@Rollback(false)
    def parallelSaveAndReadForJava() {

        val orderIds = ArrayBuffer[Long]()

        JpaParallels.runAction(emf, (0 to entityCount).toIterable,
            new JpaParRunnable[Int] {
                override def run(em: EntityManager, x: Int): Unit = {
                    val order = createOrder(x)
                    em.persist(order)
                    orderIds += order.id
                }
            })

        Thread.sleep(10)

        var orders =
            JpaParallels.callFunc(emf, orderIds,
                new JpaParCallable[Long, ParallelOrder] {
                    override def call(em: EntityManager, id: Long): ParallelOrder = {
                        val order = em.find(classOf[ParallelOrder], id)
                        Hibernate.initialize(order.items)
                        em.detach(order)
                        order
                    }
                })

        assert(orders.forall(x => x != null))

        // 여기서부터는 hibernate-redis second 캐시에서도 잘 되는지 알아보기 위함입니다.
        orders =
            JpaParallels.callFunc(emf, orderIds,
                new JpaParCallable[Long, ParallelOrder] {
                    override def call(em: EntityManager, id: Long): ParallelOrder = {
                        val order = em.find(classOf[ParallelOrder], id)
                        Hibernate.initialize(order.items)
                        em.detach(order)
                        order
                    }
                })

        assert(orders.forall(x => x != null))

        orders =
            JpaParallels.callFunc(emf, orderIds,
                new JpaParCallable[Long, ParallelOrder] {
                    override def call(em: EntityManager, id: Long): ParallelOrder = {
                        val order = em.find(classOf[ParallelOrder], id)
                        Hibernate.initialize(order.items)
                        em.detach(order)
                        order
                    }
                })

        assert(orders.forall(x => x != null))

    }

    private def createOrder(x: Long) = {
        val order = new ParallelOrder()
        order.no = "order no. " + x

        val item1 = new ParallelOrderItem()
        item1.name = "Item1-" + x
        item1.order = order
        order.items.add(item1)

        val item2 = new ParallelOrderItem()
        item2.name = "Item2-" + x
        item2.order = order
        order.items.add(item2)

        order
    }
}

@Entity
@hba.Cache(region = "parallel", usage = hba.CacheConcurrencyStrategy.READ_WRITE)
@hba.DynamicInsert
@hba.DynamicUpdate
class ParallelOrder extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    @Column(name = "orderId")
    var id: lang.Long = _

    def getId = id

    var no: String = _

    @OneToMany(mappedBy = "order", cascade = Array(CascadeType.ALL), orphanRemoval = true)
    val items: util.List[ParallelOrderItem] = new util.ArrayList[ParallelOrderItem]

    @inline
    override def hashCode(): Int = Hashs.compute(no)

    @inline
    override protected def buildStringHelper: ToStringHelper =
        super.buildStringHelper
        .add("no", no)
}

@Entity
@hba.Cache(region = "parallel", usage = hba.CacheConcurrencyStrategy.READ_WRITE)
@hba.DynamicInsert
@hba.DynamicUpdate
class ParallelOrderItem extends HibernateEntity[lang.Long] {

    @Id
    @GeneratedValue
    @Column(name = "orderItemId")
    var id: lang.Long = _

    def getId = id

    @ManyToOne
    @JoinColumn(name = "orderId")
    var order: ParallelOrder = _

    var name: String = _

    @inline
    override def hashCode(): Int = Hashs.compute(name)
}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/parallel/JpaParallelsTest.java
package debop4s.data.orm.jtests.mapping.parallel;

import debop4s.core.collections.NumberRange;
import debop4s.data.orm.jpa.utils.JpaParCallable;
import debop4s.data.orm.jpa.utils.JpaParallels;
import debop4s.data.orm.jtests.jpa.JpaTestBase;
import lombok.extern.slf4j.Slf4j;
import org.hibernate.Hibernate;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import org.modelmapper.internal.util.Lists;
import org.springframework.test.annotation.Rollback;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.ArrayList;
import java.util.List;

import static org.fest.assertions.Assertions.assertThat;

@Slf4j
@Transactional
@Ignore("현재 테스트 중입니다. 멀티스레드로 JPA 작업은 하지 마시기 바랍니다.")
public class JpaParallelsTest extends JpaTestBase {

    @PersistenceContext EntityManager em;
    private static final long entityCount = 10;

    @Before
    public void before() {
        em.createQuery("delete from ParallelOrderItem").executeUpdate();
        em.createQuery("delete from ParallelOrder").executeUpdate();
        em.flush();
        em.clear();
    }

    @Test
    @Transactional(propagation = Propagation.REQUIRES_NEW)
    @Rollback(false)
    public void parallelSaveAndRead() {
        final List<Long> orderIds = new ArrayList<Long>();

        final List<Long> orderNos = Lists.from(NumberRange.range(entityCount).iterator());

        for (long i = 0; i < entityCount; i++) {
            ParallelOrder order = createOrder(i);
            order.setId(i);
            em.persist(order);
            orderIds.add(i);
        }
        em.flush();
        em.clear();

        log.debug("orderIds=[{}]", orderIds);

        Iterable<ParallelOrder> orders =
                JpaParallels.callFunc(emf, orderIds, new JpaParCallable<Long, ParallelOrder>() {
                    @Override
                    public ParallelOrder call(EntityManager em, Long elem) {
                        log.debug("load order. id=[{}]", elem);
                        ParallelOrder order = em.find(ParallelOrder.class, elem);
                        Hibernate.initialize(order.getItems());
                        em.detach(order);
                        return order;
                    }
                });
        for (ParallelOrder order : orders) {
            assertThat(order).isNotNull();
        }

        // read from second cache
        orders = JpaParallels.callFunc(emf, orderIds, new JpaParCallable<Long, ParallelOrder>() {
            @Override
            public ParallelOrder call(EntityManager em, Long elem) {
                ParallelOrder order = em.find(ParallelOrder.class, elem);
                Hibernate.initialize(order.getItems());
                return order;
            }
        });
        for (ParallelOrder order : orders) {
            assertThat(order).isNotNull();
        }

        orders = JpaParallels.callFunc(emf, orderIds, new JpaParCallable<Long, ParallelOrder>() {
            @Override
            public ParallelOrder call(EntityManager em, Long elem) {
                ParallelOrder order = em.find(ParallelOrder.class, elem);
                Hibernate.initialize(order.getItems());
                return order;
            }
        });
        for (ParallelOrder order : orders) {
            assertThat(order).isNotNull();
        }

    }

    private ParallelOrder createOrder(Long orderNo) {
        ParallelOrder order = new ParallelOrder();
        order.setNo("order no." + orderNo);

        ParallelOrderItem item1 = new ParallelOrderItem();
        item1.setName("Item1-" + orderNo);
        item1.setOrder(order);
        order.getItems().add(item1);

        ParallelOrderItem item2 = new ParallelOrderItem();
        item2.setName("Item1-" + orderNo);
        item2.setOrder(order);
        order.getItems().add(item2);

        return order;
    }
}
